### PR TITLE
implement the scanning function of BLE 

### PR DESCRIPTION
**Pull request by Liuguimei**
1.show every single scan device
2.Click item in the list will switch to the detail page
3.Connect to a single BLE device

### CODE REVIEW

File: app/src/main/java/com/tymphay/tymwork/ui/ConnectBluetoothActivity.kt [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    private fun initView(){
+        supportActionBar?.apply {
+            title = "Connection"
+            setDisplayHomeAsUpEnabled(true)
+        }
+        //获取从上个页面传递过来的设备
+        val device = intent.getParcelableExtra<BluetoothDevice>("device")
+        //将设备详细信息显示在UI上
+        Log.v("Detail", "${device?.name}")
+        tv_device_name_1.setText("设备名称：${device?.name}")
```
[the remark this answers is not in the document]
Comment (Liuguimei) [reply app/src/main/java/com/tymphay/tymwork/ui/ConnectBluetoothActivity.kt]: resolved

File: app/build.gradle [hunk 2]
```diff
@@ -42,7 +47,10 @@ dependencies {
     androidTestImplementation 'androidx.test.ext:junit:1.1.3'
     androidTestImplementation 'androidx.test.espresso:espresso-core:3.4.0'
 
-    //Navigation
-    implementation "androidx.navigation:navigation-fragment-ktx:2.4.2"
-    implementation "androidx.navigation:navigation-ui-ktx:2.4.2"
+    //蓝牙扫描库
+    implementation 'no.nordicsemi.android.support.v18:scanner:1.5.0'
```
[the remark this answers is not in the document]
Comment (Liuguimei) [reply app/build.gradle:47]: resolved

File: app/src/main/java/com/tymphay/tymwork/adapter/ConnectDeviceAdapter.kt [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+        //获取蓝牙设备名称
+        holder.deviceName?.text= list?.get(position)?.name
+    }
+
+    override fun getItemCount(): Int {
+        return list!!.size
+    }
+}
+
+class DeviceViewHolder(itemView: View): RecyclerView.ViewHolder(itemView) {
```
[the remark this answers is not in the document]
Comment (Liuguimei) [reply app/src/main/java/com/tymphay/tymwork/adapter/ConnectDeviceAdapter.kt:28]: resolved

File: app/src/main/java/com/tymphay/tymwork/ui/MainActivity.kt [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+        initView()
+    }
+
+    //检查Android版本
+    private fun checkAndroidVersion(){
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M){
+            //判断权限是否开启
+            when (ContextCompat.checkSelfPermission(this,Manifest.permission.ACCESS_COARSE_LOCATION)) {
+                PackageManager.PERMISSION_GRANTED -> openBluetooth()    //权限已打开
+                else -> requestPermissions(arrayOf(Manifest.permission.ACCESS_COARSE_LOCATION), 1)  //权限未打开,请求权限
```
[the remark this answers is not in the document]
Comment (Liuguimei) [reply app/src/main/java/com/tymphay/tymwork/ui/MainActivity.kt:68]: resolved

File: app/src/main/java/com/tymphay/tymwork/bean/ConnectDevice.kt [hunk 5]
```diff
@@ -0,0 +1,6 @@
+package com.tymphay.tymwork.bean
+
+import android.bluetooth.BluetoothDevice
+
+//存放已连接设备结果
+class ConnectDevice(var device: BluetoothDevice, var name: String)
```
[the remark this answers is not in the document]
Comment (Liuguimei) [reply app/src/main/java/com/tymphay/tymwork/bean/ConnectDevice.kt:6]: resolved

File: app/src/main/java/com/tymphay/tymwork/ui/ConnectBluetoothActivity.kt [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+import kotlin.collections.HashMap
+
+class ConnectBluetoothActivity :AppCompatActivity() {
+
+    //状态缓存
+    private var stringBuffer= StringBuffer()
+    //连接设备适配器
+    var connectDeviceAdapter : ConnectDeviceAdapter? =null
+    //Key:service name      Value:service uuid
+    val uuids : HashMap<String,String> = HashMap<String,String>()
```
[the remark this answers is not in the document]
Comment (Liuguimei) [reply app/src/main/java/com/tymphay/tymwork/ui/ConnectBluetoothActivity.kt:28]: resolved

File: app/src/main/java/com/tymphay/tymwork/ui/ConnectBluetoothActivity.kt [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+import java.util.*
+import kotlin.collections.ArrayList
+import kotlin.collections.HashMap
+
+class ConnectBluetoothActivity :AppCompatActivity() {
+
+    //状态缓存
+    private var stringBuffer= StringBuffer()
+    //连接设备适配器
+    var connectDeviceAdapter : ConnectDeviceAdapter? =null
```
[the remark this answers is not in the document]
Comment (Liuguimei) [reply app/src/main/java/com/tymphay/tymwork/ui/ConnectBluetoothActivity.kt:26]: resolved

File: app/src/main/java/com/tymphay/tymwork/adapter/ScanListAdapter.kt [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+            onItemClick.invoke(position)
+        }
+    }
+
+    override fun getItemCount(): Int {
+        return list!!.size
+    }
+}
+
+class DeviceViewHolderScan(itemView: View): RecyclerView.ViewHolder(itemView) {
```
[the remark this answers is not in the document]
Comment (Liuguimei) [reply app/src/main/java/com/tymphay/tymwork/adapter/ScanListAdapter.kt:48]: resolved

File: app/src/main/java/com/tymphay/tymwork/ui/ConnectBluetoothActivity.kt [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+    val uuids : HashMap<String,String> = HashMap<String,String>()
+    lateinit var gatt: BluetoothGatt
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_connect_bluetooth)
+
+        //service uuid:
+        uuids.put("00001800-0000-1000-8000-00805f9b34fb","Generic Access Profile")
+//        uuids.put("00001801-0000-1000-8000-00805f9b34fb","Generic Attribute Profile")
```
[the remark this answers is not in the document]
Comment (Liuguimei) [reply app/src/main/java/com/tymphay/tymwork/ui/ConnectBluetoothActivity.kt:37]: resolved

File: app/src/main/java/com/tymphay/tymwork/callback/BleCallback.kt [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+        uiCallback.state(
+            when (newState) {
+                BluetoothProfile.STATE_CONNECTED -> {   //连接成功
+
+                    //获取MtuSize,触发onMtuChanged回调
+                   // gatt.requestMtu(512)
+
+                    //连接成功后，发现服务,触发onServicesDiscovered回调
+                    gatt.discoverServices()
+                    "连接成功"
```
[the remark this answers is not in the document]
Comment (Liuguimei) [reply app/src/main/java/com/tymphay/tymwork/callback/BleCallback.kt]: resolved

File: app/src/main/java/com/tymphay/tymwork/view/ConnectBluetoothActivity.kt [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+import com.tymphay.tymwork.adapter.ConnectDeviceAdapter
+import com.tymphay.tymwork.model.ConnectDevice
+import kotlinx.android.synthetic.main.activity_connect_bluetooth.*
+import kotlin.collections.ArrayList
+import kotlin.collections.HashMap
+
+class ConnectBluetoothActivity :AppCompatActivity() {
+
+    //状态缓存
+    var stringBuffer= StringBuffer()
```
Comment (rdaoshiken) [inline app/src/main/java/com/tymphay/tymwork/view/ConnectBluetoothActivity.kt:20]: private

File: app/src/main/AndroidManifest.xml [hunk 12]
```diff
@@ -2,13 +2,27 @@
...
+    <!--蓝牙通讯权限,用来发现或操作蓝牙设备-->
+    <uses-permission android:name="android.permission.BLUETOOTH_ADMIN" />
+    <!-- 支持ble的设备 -->
+    <uses-feature
+        android:name="android.hardware.bluetooth_le"
+        android:required="true" />
+    <uses-permission android:name="android.permission.ACCESS_COARSE_LOCATION"/>
+    <uses-permission android:name="android.permission.ACCESS_FINE_LOCATION"/>
+    <uses-permission android:name="android.permission.BLUETOOTH_CONNECT" />
+
```
Comment (rdaoshiken) [inline app/src/main/AndroidManifest.xml:16]: android 12 ble permission

File: app/src/main/java/com/tymphay/tymwork/view/ConnectBluetoothActivity.kt (hunk 11 above)
Comment (Liuguimei) [reply app/src/main/java/com/tymphay/tymwork/view/ConnectBluetoothActivity.kt:23]: resolved

**Review by rdaoshiken — APPROVED** [review]

**Pull request merged**